src/awesome/class.rs

```diff
@@ -0,0 +1,163 @@
+        meta.set(name, meth)?;
+        Ok(self)
+    }
+
+    pub fn property(self, prop: Property<'lua>) -> rlua::Result<Self> {
+        let properties = self.class.table.get::<_, Table>("properties")?;
+        let length = properties.len().unwrap_or(0) + 1;
+        properties.set(length, prop)?;
+        Ok(self)
+    }
+
+    pub fn save_class(mut self, name: &str)
+                      -> rlua::Result<Self> {
+        self.lua.globals().set(name, self.class.table)?;
+        self.class.table = self.lua.globals().get(name)?;
+        Ok(self)
+    }
+
+    pub fn build(self) -> rlua::Result<Class<'lua>> {
+        Ok(self.class)
+    }
+}
+
+impl <'lua> ToLua<'lua> for Class<'lua> {
+    fn to_lua(self, lua: &'lua Lua) -> rlua::Result<Value<'lua>> {
+        self.table.to_lua(lua)
+    }
+}
+
+impl Default for ClassState {
+    fn default() -> Self {
+        ClassState {
+            name: String::default(),
+            signals_global: Option::default(),
+            parent: Option::default(),
+            allocator: Option::default(),
+            collector: Option::default(),
+            checker: Option::default(),
+            instances: 0,
+            index_miss_handler_global: Option::default(),
+            newindex_miss_handler_global: Option::default(),
+
+        }
+    }
+}
+
+impl UserData for ClassState {}
+
+impl <'lua> Class<'lua> {
+    pub fn new(lua: &'lua Lua,
+               allocator: Option<Allocator>,
+               collector: Option<Collector>,
+               checker: Option<Checker>)
+               -> rlua::Result<ClassBuilder<'lua>> {
+        let mut class = ClassState::default();
+        class.allocator = allocator;
+        class.collector = collector;
+        class.checker = checker;
+        let table = lua.create_table();
+        table.set("data", class)?;
+        table.set("properties", Vec::<Property>::new().to_lua(lua)?)?;
+        let meta = lua.create_table();
+        meta.set("signals", lua.create_table())?;
+        meta.set("set_index_miss_handler",
+                 lua.create_function(set_index_miss_handler).bind(table.clone())?)?;
+        meta.set("set_newindex_miss_handler",
+                 lua.create_function(set_newindex_miss_handler).bind(table.clone())?)?;
+        meta.set("__index", meta.clone())?;
+        table.set_metatable(Some(meta));
+        Ok(ClassBuilder{
+            lua: lua,
+            class: Class { table }
+        })
+    }
+
+    #[allow(dead_code)]
+    pub fn properties(&self) -> rlua::Result<Table<'lua>> {
+        self.table.get("properties")
+    }
+
+    pub fn checker(&self) -> rlua::Result<Option<Checker>> {
+        self.table.get::<_, ClassState>("data")
+            .map(|state| state.checker)
+    }
+}
+
+impl <'lua> From<Table<'lua>> for Class<'lua> {
+    fn from(table: Table<'lua>) -> Self {
+        Class { table }
+    }
+}
+
+fn set_index_miss_handler<'lua>(_: &'lua Lua, (obj, func): (Table, Function))
```

Comment:
Is this for the class? Make it a method on the class, since we don't use lua (or does it need this sig? If so, still make it a method please)